src/coreclr/tools/Common/TypeSystem/Common/MetadataVirtualMethodAlgorithm.cs

```diff
@@ -826,5 +831,125 @@ public static IEnumerable<MethodDesc> EnumAllVirtualSlots(MetadataType type)
+                        MethodDesc runtimeInterfaceMethod = TryResolveInterfaceMethodOnVariantCompatibleInterface(runtimeInterfaceType, interfaceMethod);
+
+                        if (runtimeInterfaceMethod != null)
+                        {
+                            resolvedMethodOnType = TryResolveVirtualStaticMethodOnThisType(typeToCheck, runtimeInterfaceMethod);
+                        }
+                        if (resolvedMethodOnType != null)
+                        {
+                            return resolvedMethodOnType;
+                        }
+                    }
+                }
+            }
+            return null;
+        }
+
+        private static MethodDesc TryResolveInterfaceMethodOnVariantCompatibleInterface(TypeDesc compatibleInterfaceType, MethodDesc interfaceMethod)
+        {
+            Debug.Assert(compatibleInterfaceType.CanCastTo(interfaceMethod.OwningType));
+            if (compatibleInterfaceType is MetadataType mdType)
+            {
+                foreach (MethodDesc runtimeInterfaceMethod in mdType.GetVirtualMethods())
+                {
+                    if (runtimeInterfaceMethod.Name == interfaceMethod.Name &&
+                        runtimeInterfaceMethod.Signature == interfaceMethod.Signature)
+                    {
+                        return runtimeInterfaceMethod;
+                    }
+                }
+            }
+            return null;
+        }
+
+        /// <summary>
+        /// Try to resolve a given virtual static interface method on a given constrained type and return the resolved method or null when not found.
+        /// </summary>
+        /// <param name="constrainedType">Type to attempt method resolution on</param>
+        /// <param name="interfaceType">Interface declaring the method</param>
+        /// <param name="interfaceMethod">Method to resolve</param>
+        /// <returns>MethodDesc of the resolved method or null when not found (runtime lookup must be used)</returns>
+        private static MethodDesc TryResolveVirtualStaticMethodOnThisType(TypeDesc constrainedType, MethodDesc interfaceMethod)
+        {
+            if (constrainedType is MetadataType mdType)
+            {
+                foreach (MethodImplRecord methodImpl in mdType.FindMethodsImplWithMatchingDeclName(interfaceMethod.Name) ?? Array.Empty<MethodImplRecord>())
+                {
+                    if (methodImpl.Decl == interfaceMethod)
+                    {
+                        MethodDesc resolvedMethodImpl = methodImpl.Body;
+                        if (resolvedMethodImpl.OwningType != mdType)
+                        {
+                            ThrowHelper.ThrowMissingMethodException(constrainedType, resolvedMethodImpl.Name, resolvedMethodImpl.Signature);
+                        }
+                        if (interfaceMethod.HasInstantiation || methodImpl.Body.HasInstantiation || constrainedType.HasInstantiation)
```

Comment:
```suggestion
                        if (interfaceMethod.HasInstantiation)
```

If `interfaceMethod.HasInstantiation != methodImpl.Body.HasInstantiation`, terrible things will happen below, so we might as well assume that and consider the Body check redundant.

I believe InstantiateSignature is going to be a no-op if the constrained type has an instantiation because that part is already instantiated.

Reply:
Fixed in the latest commit, thanks for the suggestion.